src/Runtime/Distributions/DiscreteChar.cs

```diff
@@ -2122,172 +1465,694 @@ public static IEnumerable<CharRangePair> CombinedRanges(DiscreteChar distributio
+                            this.symbolRepresentation = WordCharSymbolRepresentation;
+                        }
+                        // TODO: handle uniform
+                        else
+                        {
+                            var representation = new StringBuilder();
+                            this.AppendRanges(representation, false);
+                            this.symbolRepresentation = representation.ToString();
+                        }
+                    }
+
+                    return symbolRepresentation;
+                }
+            }
+
+            private void AppendRanges(StringBuilder representation, bool escape)
+            {
+                if (this.Ranges.Length >= 1)
+                {
+                    representation.Append('[');
+                    foreach (var range in this.Ranges)
+                    {
+                        AppendCharacterRange(representation, range, escape);
+                    }
+
+                    representation.Append(']');
+                }
+            }
+
+            #endregion
+        }
+
+        /// <summary>
+        /// Internal cache for the most used character distributions. Creates them lazily on demand
+        /// </summary>
+        internal sealed class StorageCache
+        {
+            public static Lazy<Storage> Uniform = new Lazy<Storage>(
+                () => Storage.CreateUncached(new CharRange[] { }, UniformProb, CharClasses.Uniform, UniformRegexRepresentation));
+
+            public static Lazy<Storage> Digit = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges("09", CharClasses.Digit, DigitRegexRepresentation));
+
+            public static Lazy<Storage> Lower = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges(LowerCaseCharacterRanges, CharClasses.Lower, LowerRegexRepresentation));
+
+            public static Lazy<Storage> Upper = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges(UpperCaseCharacterRanges, CharClasses.Upper, UpperRegexRepresentation));
+
+            public static Lazy<Storage> Letter = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges(LetterCharacterRanges, CharClasses.Letter, LetterRegexRepresentation));
+
+            public static Lazy<Storage> LetterOrDigit = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges(LetterCharacterRanges + "09", CharClasses.LetterOrDigit, LetterOrDigitRegexRepresentation));
+
+            public static Lazy<Storage> WordChar = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges(LetterCharacterRanges + "09__", CharClasses.WordChar, WordCharRegexRepresentation));
+
+            public static Lazy<Storage> NonWordChar = new Lazy<Storage>(
+                () => WordChar.Value.Complement());
+
+            public static Lazy<Storage> Whitespace = new Lazy<Storage>(
+                () => Storage.CreateUniformInRanges("\t\r  ", CharClasses.Unknown, null));
+
```

Comment:
Might as well materialize them immediately and get rid of the call through Lazy.

Reply:
Reasonable.

I thought about that, but than decided that will bring to live 10 or so objects even in programs which don't need them. Given that in our project we mostly deal with string distributions, I don't mind going from lazy materialization to eager - we are going to use these objects anyway.

I would love to hear from project maintainers like @tminka, @yordanz or @a-vishar  what is preferred. Unless answer is given, I will go for eager materialization.

Reply:
UPD. Actually lazy is not required due to way static constructors work